server/src/main/java/org/candlepin/controller/refresher/visitors/ProductNodeVisitor.java

```diff
@@ -0,0 +1,340 @@
+        if (productCurator == null) {
+            throw new IllegalArgumentException("productCurator is null");
+        }
+
+        if (ownerProductCurator == null) {
+            throw new IllegalArgumentException("ownerProductCurator is null");
+        }
+
+        this.productCurator = productCurator;
+        this.ownerProductCurator = ownerProductCurator;
+
+        this.ownerProductEntities = new HashSet<>();
+        this.ownerProductUuidMap = new HashMap<>();
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public Class<Product> getEntityClass() {
+        return Product.class;
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public void processNode(NodeProcessor processor, NodeMapper mapper,
+        EntityNode<Product, ProductInfo> node) {
+
+        boolean childrenUpdated = false;
+        boolean nodeChanged = false;
+
+        if (node.visited()) {
+            return;
+        }
+
+        // Process children nodes first (depth-first), so we can update references and avoid
+        // rework; also check if we need to make reference updates on this entity.
+        for (EntityNode child : (Set<EntityNode>) node.getChildrenNodes()) {
+            processor.processNode(child);
+            childrenUpdated |= child.changed();
+        }
+
+        Product existingEntity = node.getExistingEntity();
+        ProductInfo importedEntity = node.getImportedEntity();
+
+        if (existingEntity != null) {
+            if (importedEntity != null) {
+                nodeChanged = ProductManager.isChangedBy(existingEntity, importedEntity);
+            }
+
+            if (nodeChanged || childrenUpdated) {
+                Product updatedEntity = this.createEntity(mapper, node);
+                node.setUpdatedEntity(updatedEntity);
+
+                // Store the mapping to be updated later
+                Map<String, String> productUuidMap = this.ownerProductUuidMap.get(node.getOwner());
+                if (productUuidMap == null) {
+                    productUuidMap = new HashMap<>();
+                    this.ownerProductUuidMap.put(node.getOwner(), productUuidMap);
+                }
+                productUuidMap.put(existingEntity.getUuid(), updatedEntity.getUuid());
+
+                node.markChanged();
+            }
+        }
+        else {
+            // Node is new
+            Product updatedEntity = this.createEntity(mapper, node);
+            node.setUpdatedEntity(updatedEntity);
+
+            // Create a new owner-product mapping for this entity. This will get persisted
+            // later during completion
+            this.ownerProductEntities.add(new OwnerProduct(node.getOwner(), updatedEntity));
+
+            node.markChanged();
+        }
+
+        node.markVisited();
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public void complete() {
+        this.ownerProductCurator.saveAll(this.ownerProductEntities, false, true);
```

Comment:
The ProductManager did flush when saving the core products, then again when saving the owner-products. Here it seems we're deferring flushing for when? Shouldn't we flush before CandlepinPoolManager starts processing the pools? (same question applies for `ContentNodeVisitor.complete`)

Reply:
yes, evict to false would be more clear

Reply:
Done.